### PR TITLE
Attach SnowFlake line comments to SELECT statements

### PR DESCRIPTION
**Pull request by ericvergnaud**
There is a need to transpile SQL comments
This PR:
 - sends `LINE_COMMENT`s to a dedicated ANLTR4 channel, such that comments do not interfere with SQL constructs
 - adds a `comments` array field to the `LogicalPlan` case class
 - moves the `origin` field of `TreeNode` to a 2nd parameter list such that it is ignored during comparisons, and refactors accordingly
 - refactors the `LogicalPlan` pipeline such that it's possible to use the original `TokenStream` to fetch the comment tokens and apply them appropriately to the plan items. The abstract plan is also no loger dependent on ANTLR4.
 - adds a test for checking transpilation

Progresses #869 

Future PRs will expand this to:
 - TSQL parser
 - more constructs
 - inline comments

### CODE REVIEW

File: core/src/main/scala/com/databricks/labs/remorph/intermediate/plans.scala [hunk 1]
```diff
@@ -6,7 +6,7 @@ package com.databricks.labs.remorph.intermediate
  * the [[Command]] type that is used to execute commands on the server. [[Plan]] is a union of Spark's LogicalPlan and
  * QueryPlan.
  */
-abstract class Plan[PlanType <: Plan[PlanType]] extends TreeNode[PlanType] {
+abstract class Plan[PlanType <: Plan[PlanType]]()(origin: Origin) extends TreeNode[PlanType]()(origin) {
```
Comment (ericvergnaud) [inline core/src/main/scala/com/databricks/labs/remorph/intermediate/plans.scala:9]: origin in 2nd parameter list is required to only compare relevant fields

File: core/src/main/scala/com/databricks/labs/remorph/parsers/snowflake/SnowflakePlanParser.scala [hunk 2]
```diff
@@ -1,20 +1,68 @@
...
+class SnowflakePlanParser extends PlanParser {
 
   private val vc = new SnowflakeVisitorCoordinator(SnowflakeParser.VOCABULARY, SnowflakeParser.ruleNames)
 
-  override protected def createLexer(input: CharStream): Lexer = new SnowflakeLexer(input)
-  override protected def createParser(stream: TokenStream): SnowflakeParser = new SnowflakeParser(stream)
-  override protected def createTree(parser: SnowflakeParser): ParserRuleContext = parser.snowflakeFile()
-  override protected def createPlan(tree: ParserRuleContext): ir.LogicalPlan = vc.astBuilder.visit(tree)
-  override protected def addErrorStrategy(parser: SnowflakeParser): Unit =
+  override def parseLogicalPlan(parsing: Parsing): Transformation[LogicalPlan] = {
```
Comment (ericvergnaud) [inline core/src/main/scala/com/databricks/labs/remorph/parsers/snowflake/SnowflakePlanParser.scala:16]: we need both the plan and the token stream to attach comments so had to move away from the original design

File: core/src/main/scala/com/databricks/labs/remorph/parsers/tsql/TSqlPlanParser.scala [hunk 3]
```diff
@@ -1,19 +1,51 @@
...
+class TSqlPlanParser extends PlanParser {
 
   val vc = new TSqlVisitorCoordinator(TSqlParser.VOCABULARY, TSqlParser.ruleNames)
 
-  override protected def createLexer(input: CharStream): Lexer = new TSqlLexer(input)
-  override protected def createParser(stream: TokenStream): TSqlParser = new TSqlParser(stream)
-  override protected def createTree(parser: TSqlParser): ParserRuleContext = parser.tSqlFile()
-  override protected def createPlan(tree: ParserRuleContext): ir.LogicalPlan = vc.astBuilder.visit(tree)
-  override protected def addErrorStrategy(parser: TSqlParser): Unit = parser.setErrorHandler(new TSqlErrorStrategy)
+  override def parseLogicalPlan(parsing: Parsing): Transformation[LogicalPlan] = {
```
Comment (ericvergnaud) [inline core/src/main/scala/com/databricks/labs/remorph/parsers/tsql/TSqlPlanParser.scala:16]: we need both the plan and the token stream to attach comments so had to move away from the original design

File: core/src/main/scala/com/databricks/labs/remorph/transpilers/Transpiler.scala [hunk 4]
```diff
@@ -49,14 +48,12 @@ trait Formatter {
 
 abstract class BaseTranspiler extends Transpiler with Formatter with TransformationConstructors {
 
-  protected val planParser: PlanParser[_]
+  protected val planParser: PlanParser
```
Comment (ericvergnaud) [inline core/src/main/scala/com/databricks/labs/remorph/transpilers/Transpiler.scala:51]: drop dependency on ANTLR$ - no longer required

File: core/src/main/scala/com/databricks/labs/remorph/intermediate/trees.scala [hunk 5]
```diff
@@ -252,31 +265,12 @@ abstract class TreeNode[BaseType <: TreeNode[BaseType]] extends Product {
...
-          // to older heuristic
-          false
-        } else {
-          val argsArray: Array[Class[_]] = allArgs.map(_.getClass)
-          isAssignable(argsArray, ctor.getParameterTypes)
-        }
-      }
-      .getOrElse(ctors.maxBy(_.getParameterTypes.length)) // fall back to older heuristic
+
+    val (defaultCtor, allArgs) = locateConstructor(ctors, newArgs, appendOrigin = false)
```
Comment (ericvergnaud) [inline core/src/main/scala/com/databricks/labs/remorph/intermediate/trees.scala:269]: the constructor may or not require the Origin paranter from the 2nd parameter list, so try locating constructor with and without it

File: core/src/test/scala/com/databricks/labs/remorph/discovery/AnonymizerTest.scala [hunk 6]
```diff
@@ -62,7 +62,7 @@ class AnonymizerTest extends AnyWordSpec with Matchers {
         Fingerprint(
           "id",
           new Timestamp(1725032011000L),
-          "290f4d72ca8faeb28873d8fff779ce93ed5cdb69",
+          "93f60d8795c8bffa2aafe174ae8a867b42235755",
```
Comment (ericvergnaud) [inline core/src/test/scala/com/databricks/labs/remorph/discovery/AnonymizerTest.scala:65]: To be discussed: should we generate a plan from a failed parse

File: core/src/main/scala/com/databricks/labs/remorph/intermediate/plans.scala (hunk 1 above)
Comment (vil1) [reply core/src/main/scala/com/databricks/labs/remorph/intermediate/plans.scala:9]: As per [the guidelines](https://github.com/databricks/scala-style-guide?tab=readme-ov-file#multi-param-list) we should avoid multiple parameter lists. Could we come up with a solution that doesn't involve adding these parameter lists?

Comment (ericvergnaud) [reply core/src/main/scala/com/databricks/labs/remorph/intermediate/plans.scala:9]: I agree that mpl _should_ be `avoided`, but that's the only official way I know to exclude origin from comparisons in both TreeNode and all derived classes. I'm happy to consider any suggestion.

Comment (vil1) [reply core/src/main/scala/com/databricks/labs/remorph/intermediate/plans.scala:9]: From the back of my head, I guess something along the lines of:

```scala
trait TreeNode {
  private [this] var _origin: Origin = Origin.empty

  def origin: Origin = _origin

  def withOrigin(updatedOrigin: Origin): this.type = {
    _origin = updatedOrigin
    this 
  }
}
```

should do the trick

**Comment by ericvergnaud** [comment]
> This is quite a big change and somewhat challenging to review. Could you try and split this PR into smaller ones?
> 
> I believe we could at least have the changes related to `Origin` in a separate PR.
> 
> Stacking PRs is fine (PR `n` targeting the branch of PR `n - 1`).

ok, see https://github.com/databrickslabs/remorph/pull/1189

File: core/src/test/scala/com/databricks/labs/remorph/discovery/AnonymizerTest.scala (hunk 6 above)
Comment (jimidle) [reply core/src/test/scala/com/databricks/labs/remorph/discovery/AnonymizerTest.scala:65]: Yes. Parse failures are generated as comments.

File: core/src/main/scala/com/databricks/labs/remorph/parsers/snowflake/SnowflakePlanParser.scala (hunk 2 above)
Comment (jimidle) [reply core/src/main/scala/com/databricks/labs/remorph/parsers/snowflake/SnowflakePlanParser.scala:16]: > we need both the plan and the token stream to attach comments so had to move away from the original single-enity-transform design

Could we not just accumulate he comments in a structure that allows search by their location? Either after the lexer is called or just within the lexer directly? Then we would not need the token stream and this map or whatever it is can be retrieved from the common lexer, then passed to the generator. 

Perhaps we need to settle on a design for how we are resolving comment location before coding?

**Comment by ericvergnaud** [comment]
superseded by multiple smaller PRs

**Pull request closed without merging**